lib/reference-resolver.js

```diff
@@ -0,0 +1,134 @@
+   * });
+   *
+   * // Resolve a patient contained in someResource (see:
+   * // http://hl7.org/fhir/STU3/references.html#contained)
+   * client.resolve("#patient-1", someResource).then((patient) => {
+   *   console.log(patient);
+   * });
+   *
+   * @param {String} reference the FHIR reference
+   * @param {Object} [context] bundle or FHIR resource, optional
+   *
+   * @return {Promise<Object>} FHIR resource
+   */
+  async resolve(reference, context) {
+    if (context === undefined) {
+      if (reference.startsWith('http')) {
+        return this.resolveAbsoluteReference(reference);
+      }
+      return this.client.httpClient.get(reference);
+    }
+    if (reference.startsWith('#')) {
+      return this.resolveContainedReference(reference, context);
+    }
+    return this.resolveBundleReference(reference, context);
+  }
+
+  /**
+   * Private function for Client to resolve absolute references
+   *
+   * @private
+   *
+   * @async
+   *
+   * @param {String} reference the FHIR reference
+   *
+   * @return {Promise<Object>} FHIR resource
+   */
+  async resolveAbsoluteReference(reference) {
+    if (reference.startsWith(this.client.baseUrl)) {
+      return this.client.httpClient.get(reference);
+    }
+    const { baseUrl, resourceType, id } = splitReference(reference);
+    const { Client } = require('./client'); // eslint-disable-line global-require
+    return (new Client({ baseUrl })).read({ resourceType, id });
+  }
+
+  /**
+   * Private function for Client to resolve contained references
+   *
+   * @private
+   *
+   * @param {String} reference the FHIR reference
+   * @param {Object} [context] FHIR resource
+   *
+   * @return {Object} FHIR resource
+   * @throws Will throw if a contained resource which matches the reference
+   *   can't be found
+   */
+  resolveContainedReference(reference, context) { // eslint-disable-line class-methods-use-this
+    if (context.contained) {
+      const referenceId = reference.slice(1);
+      const containedResource = context.contained.find(resource => resource.id === referenceId);
+      if (containedResource) {
+        return containedResource;
+      }
+    }
+    throw (new Error(`Unable to resolve contained reference: ${reference}`));
+  }
+
+  /**
+   * Private function for Client to resolve references in a bundle. If not
+   * contained in the bundle, the resources will be fetched from a FHIR server.
+   *
+   * @private
+   *
+   * @async
+   *
+   * @param {String} reference the FHIR reference
+   * @param {Object} [context] FHIR bundle
+   *
+   * @return {Promise<Object>} FHIR resource
+   */
+  async resolveBundleReference(reference, bundle) {
+    const referenceRegEx = new RegExp(`(^|/)${reference}$`);
+    const entry = bundle.entry.find(bundleEntry => referenceRegEx.test(bundleEntry.fullUrl));
+
+    if (!entry) {
+      return this.resolve(reference);
+    }
+    return entry.resource;
+  }
+}
+
+module.exports = { ReferenceResolver };
```

Comment:
Should this just be: `module.exports = ReferenceResolver` ?

This is analogous to the "default" export, and also corresponds to the name of the file.

(despite the fact that `lib/client/.js` also does this -- perhaps it should be modified, too)

Reply:
Yeah, I agree about `Client`, I don't see a reason why it wouldn't be the default export.